### ISSUE TITLE
Unfair comparison about the bias of classifiers

### ISSUE THREAD

**Issue by 973891422** [Yutong-Dai/FedNH#1]
Hello, your work is excellent, but I have a question, for other methods, the classifier of the generic model has bias parameters,
removing the bias parameter may lead to performance degradation, which is unfair to other methods such as Fedavg?

**Comment by Yutong-Dai** [comment]
Hi there,
Thanks for reaching out. This is actually a good question.
For the discussions below, I would assume that the bias parameters refer to  `bias` argument in  `nn.Linear` class.
As an initial experiment, I also played with the `bias` terms by including and excluding them (in the last layer only). The numerical results for `FedAvg`, at least from my side,  are essentially the same (taking the randomness into account, so the accuracy metrics are close but not identical).
You can easily play around with the `bias` terms in my implementation of models, for example, [here](https://github.com/Yutong-Dai/FedNH/blob/859c03cad009159d18197713fbe151567ec4fcef/src/flbase/models/CNN.py#L19).
Should you observe any significant differences in the accuracy, please let me know.

Best,
YD

**Comment by 973891422** [comment]
Thank you for your prompt reply.
I'll give it a try and thank you again for your great work

**Comment by 973891422** [comment]
For cifar10 dirichlet0.3 convnet 100 clients and 0.1 participation ratio, I ran the code provided by your repository
Unlike the paper, the client learning rate is set to 0.1 in the configuration file, and FedNH uses orthogonal initialization

FedAvg with classifier bias
> Round:197 
 avg_test_acc: {'uniform': 0.7114999890327454, 'validclass': 0.7114999890327454, 'labeldist': 0.7114999890327454}
 pfl_avg_test_acc: {'uniform': 0.4735300064086914, 'validclass': 0.571602463722229, 'labeldist': 0.8321406364440918}
 Best test accuracy:0.711. 

FedAvg without classifier bias
> Round:197
 avg_test_acc: {'uniform': 0.7087000012397766, 'validclass': 0.7087000012397766, 'labeldist': 0.7086999416351318}
 pfl_avg_test_acc: {'uniform': 0.47092999517917633, 'validclass': 0.5682671070098877, 'labeldist': 0.8294673681259155}
 Best test accuracy:0.709.

FedNH with orthogonal initialization 
> Round:187
 avg_test_acc: {'uniform': 0.7193999886512756, 'validclass': 0.7193999886512756, 'labeldist': 0.7194000482559204}
 pfl_avg_test_acc: {'uniform': 0.5404400050640106, 'validclass': 0.6130817830562592, 'labeldist': 0.8495590448379516}
 Best test accuracy:0.719.

There is indeed some improvement, but it is not as obvious as in the paper. 
Is it due to the difference in initialization method and learning rate?

But for the weight initialization method mentioned in the paper, I couldn't find it in the code.
I'm not very good at solving optimization problems. Could you please guide me how to implement the initialization method in the paper? It's an honor to learn from your outstanding work.

Thanks again for your excellent work

**Comment by Yutong-Dai** [comment]
> Unlike the paper, the client learning rate is set to 0.1 in the configuration file, 

Good catch!  I should have updated the configuration files... Please use all hyper-parameters mentioned in the paper.

> FedNH uses orthogonal initialization. But for the weight initialization method mentioned in the paper, I couldn't find it in the code. Could you please guide me on how to implement the initialization method in the paper? 

Here is the thing, for the number reported in the paper, I solved the proposed optimization problems using the interior point method. (It can be computationally expensive if #variables are large.)  Unfortunately, I no longer have access to the server where I ran the code, so I did not release that portion of the code and provided an orthogonal initialization as a replacement. As discussed in the paper, the orthonormal basis is a feasible solution and indeed has the notion of uniformity (equidistant), but lacks the notion of maximum separation. 

I've limited bandwidth now and I will go back to re-write the code for the initialization part when I get the time.

> There is indeed some improvement, but it is not as obvious as in the paper.
Is it due to the difference in the initialization method and learning rate?

Here are multiple things going on.
1. Please change the learning rate and learning rate update strategy. 
2. Please note the variance reported in the paper. With that being said, the random seed also affects the final number you get. You could try seed 2022, which I remember was the one I've definitely used in the experiments.
3. From your provided outputs, I cannot tell whether it is from the `server_side_client` (just as an implementation hack for testing the global model's accuracy, GM reported in the paper) or an actual `client`. Plus, it is not clear what metric you are referring to. There are some tracking PFL accuracy and some tracking GM accuracy.  As a side note, `validclass`  and `'labeldist'` correspond to `PM(V)`  and `PM(L)` reported in the paper respectively. If you look at the `validclass` metric ( which I think is the correct one that should be used),  you will observe that `FedNH - 'validclass': 0.6130817830562592` and `FedAVG (with classifier bias) -  'validclass': 0.571602463722229`. Will you think it is a marginal improvement? However, I do appreciate you bringing this up, which causes your confusion. My apologies. I will revise the readme file to explain how to read the final results.
5. To address your very first question on `bias` terms, it can be seen from your provided output that, there are essentially no difference (off by ~0.3% for `validclass` and `labeldist` metrics and please remember the variance). If it still bothers you, you could add the bias term back to all appropriate competing methods and then compare.
```
FedAvg with classifier bias
pfl_avg_test_acc: {'uniform': 0.4735300064086914, 'validclass': 0.571602463722229, 'labeldist': 0.8321406364440918}
FedAvg without classifier bias
pfl_avg_test_acc: {'uniform': 0.47092999517917633, 'validclass': 0.5682671070098877, 'labeldist': 0.8294673681259155}
```


Hope I've addressed your concerns. If you have more questions, please feel free to ask. Otherwise, you could close this issue.

YD

**Comment by 973891422** [comment]
Thanks for your detailed reply, it has basically solved all my doubts.
I'm sorry for my vague description. What I mentioned before refers to the accuracy of the global model. On the personalized accuracy metric, FedNH does have a huge improvement. On the GM, maybe due to my setup, the results were slightly different, but the improvement was definitely observed.

For the inaccessibility of the interior point method, I feel like I have lost a good learning opportunity and regret it. If it is inconvenient, you can ignore it.

Finally thank you very much for your prompt and comprehensive reply, I will close this question later.

**Comment by Yutong-Dai** [comment]
I will rework and post the code for the interior point method in the near future.
Thanks for your interest in our work and your time devoted to this :-).
Happy lunar new year!
